### PR TITLE
Implement `featureInfoUrlTemplate` for Geojson data

### LINKED ISSUE #5659: 2d vector feature info bug
When working in 2D map viewer feature info will always return only one feature. In v7 it returned all features for position
v8
image

v7
image

### PR DESCRIPTION
**Pull request by zoran995**
### What this PR does

Fixes #5659, #4542, #5275

Leaflet monkey patch for onClick event is still valid, there was no major changes regarding this part. Also updated it to match the changes up to 1.7.1. This resolves selection of multiple vector features in 2D viewer.
Also implemented support for FeatureInfoMixin in 2D viewer, and connected it to GeojsonMixin.

This is also a start towards the implementation of #5136, other catalog items can be done later.

For fixing #5275 had to move the loading logic into a single `dataLoader`, when providing additional data loading method we will have to call `super.dataLoader()`. The reason for this is when using an old combination with `customDataLoader`, adding
```
const geojsonData = await this.customDataLoader();
if (isDefined(geojsonData)) return geojsonData;
```
triggered memoization somewhere which results in failed test (see #5026 and #5032 for more details). I might miss something but this is the only working solution I found.

### Checklist

-   [x] There are unit tests to verify my changes are correct or unit tests aren't applicable (if so, write quick reason why unit tests don't exist)
-   [x] I've updated CHANGES.md with what I changed.

### CODE REVIEW

File: lib/Map/geoRssConvertor.js [hunk 1]
```diff
@@ -115,10 +115,7 @@ function getGeoRssPropertiesRecursively(gmlNode, properties) {
       continue;
     }
 
-    if (
-      child.hasChildNodes() &&
-      getGeoRssPropertiesRecursively(child, properties)
-    ) {
```
Comment (zoran995) [inline lib/Map/geoRssConvertor.js]: this led to features not having any properties

File: lib/ModelMixins/GeojsonMixin.ts [hunk 2]
```diff
@@ -110,31 +118,47 @@ export default function GeoJsonMixin<
...
+     */
+    buildFeatureFromPickResult(
+      _screenPosition: Cartesian2 | undefined,
+      pickResult: any
+    ): Feature | undefined {
+      if (pickResult instanceof Entity) {
+        return Feature.fromEntityCollectionOrEntity(pickResult);
+      } else if (isDefined(pickResult?.id)) {
+        return Feature.fromEntityCollectionOrEntity(pickResult.id);
+      }
```
Comment (zoran995) [inline lib/ModelMixins/GeojsonMixin.ts]: 2D pick result is entity so we need to resolve it like that

File: lib/Models/OpenDataSoftCatalogItem.ts [hunk 3]
```diff
@@ -592,13 +593,18 @@ export default class OpenDataSoftCatalogItem
     _screenPosition: Cartesian2 | undefined,
     pickResult: any
   ) {
-    const feature = new Feature(pickResult?.id);
+    let feature: Feature | undefined;
+    if (pickResult instanceof Entity) {
+      feature = Feature.fromEntityCollectionOrEntity(pickResult);
+    } else {
+      feature = new Feature(pickResult?.id);
+    }
```
Comment (zoran995) [inline lib/Models/OpenDataSoftCatalogItem.ts]: same as for GeojsonMixin

**Comment by zoran995** [comment]
This turned into converting some catalog items to use `GeojsonMixin`, @nf-s I tried to apply changes from error handling PR

**Comment by zoran995** [comment]
This is ready for review

**Comment by zoran995** [comment]
Converted this to draft as it doesn't work on MVT layers (they are using raster feature picking), it will take some time to provide a fix

**Comment by zoran995** [comment]
This now works for MVT layers, it can also be easily  implemented for other raster layers by applying Mixin and wrapping the imagery provider (tested locally with ArcGIS map server), I just decided to skip it in this iteration and implement in follow up if this is accepted.

If a user is using cesium primitives the feature info will be shown while fetch is done in the background and after each fetch is done feature info for that feature is updated, on the other hand with raster layers fetch will be awaited and everything will be shown at the same time.

**Comment by tephenavies** [comment]
I like the idea of being able to use featureInfoUrlTemplate more widely (there's a few datasets I'd really want to use it for that are TableMixin region mapping). What's the status of this?

**Comment by zoran995** [comment]
@steve9164  This is ready for review, just needed to fix GeoRSS tests after changing it to use GeojsonMixin. Thought that it will take much more time but it was really straightforward. 
Here is a basic catalog that can be used for testing https://gist.githubusercontent.com/zoran995/8200b51b38ff6c76a954b52b223e4319/raw/6ef1e640a1732e45ca59a75baf157be6b95db0e8/feature-info-url-template-test-data.json

File: lib/Core/isDefined.ts [hunk 4]
```diff
@@ -1,3 +1,3 @@
-export default function isDefined<T>(value?: T): value is T {
-  return value !== undefined;
+export default function isDefined<T>(value: T | undefined | null): value is T {
+  return <T>value !== undefined && <T>value !== null;
```
Comment (nf-s) [inline lib/Core/isDefined.ts]: Are the type assertions here needed?

File: lib/Map/LeafletPatched.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+        layers.push(layer);
+      }
+    }
+  }
+  if (layers.length) {
+    this._fireEvent(layers, e);
+  }
+};
+
+export default L;
```
Comment (nf-s) [inline lib/Map/LeafletPatched.js]: Thanks for porting this!

File: lib/ModelMixins/FeatureInfoMixin.ts [hunk 6]
```diff
@@ -57,6 +62,17 @@ export default function FeatureInfoMixin<T extends Constructor<Target>>(
...
               });
+              // Update description of the feature after it is resolved from
+              // feature info template url
+              feature.description = new ConstantProperty(
+                describeFromProperties(
+                  feature.properties,
+                  (feature._catalogItem as any)?.currentTimeAsJulianDate ||
+                    JulianDate.now(),
+                  (feature._catalogItem as any)
+                    ?.showStringIfPropertyValueIsNull ?? false
```
Comment (nf-s) [inline lib/ModelMixins/FeatureInfoMixin.ts]:
```suggestion
                  (TimeVarying.is(this)
                    ? this.currentTimeAsJulianDate
                    : undefined) ?? JulianDate.now(),
                  TimeVarying.is(this)
                    ? this.showStringIfPropertyValueIsNull
                    : false
```

File: lib/ModelMixins/FeatureInfoMixin.ts [hunk 7]
```diff
@@ -71,7 +87,87 @@ export default function FeatureInfoMixin<T extends Constructor<Target>>(
...
+                } else {
+                  feature.properties[property] = featureInfo[property];
+                }
+              });
+              // Update description of the feature after it is resolved from
+              // feature info template url
+              feature.description = describeFromProperties(
+                feature.properties,
+                (this as any)?.currentTimeAsJulianDate || JulianDate.now(),
+                (this as any)?.showStringIfPropertyValueIsNull ?? false
```
Comment (nf-s) [inline lib/ModelMixins/FeatureInfoMixin.ts]:
```suggestion
                (TimeVarying.is(catalogItem)
                  ? catalogItem.currentTimeAsJulianDate
                  : undefined) ?? JulianDate.now(),
                TimeVarying.is(catalogItem)
                  ? catalogItem.showStringIfPropertyValueIsNull
                  : false
```

File: lib/ModelMixins/FeatureInfoMixin.ts [hunk 8]
```diff
@@ -71,7 +87,87 @@ export default function FeatureInfoMixin<T extends Constructor<Target>>(
...
+        if (
+          isDefined(catalogItem.featureInfoUrlTemplate) &&
+          isDefined(features) &&
+          features.length < catalogItem.maxRequests
+        ) {
+          for (let i = 0; i < features.length; i++) {
+            const feature = features[i];
+            const resource = new Resource({
+              url: proxyCatalogItemUrl(
+                this,
```
Comment (nf-s) [inline lib/ModelMixins/FeatureInfoMixin.ts]:
```suggestion
                catalogItem,
```

File: lib/Map/LeafletPatched.js (hunk 5 above)
Comment (nf-s) [reply lib/Map/LeafletPatched.js]: I'm currently getting this when testing locally with the "GeoJSON Test" catalog item

image

File: lib/ModelMixins/GeojsonMixin.ts [hunk 9]
```diff
@@ -794,6 +813,9 @@ function GeoJsonMixin<T extends Constructor<Model<GeoJsonTraits>>>(Base: T) {
         }
       ];
 
+      // _catalogItem property is needed for some feature picking functions (eg FeatureInfoMixin)
+      (feature as any)._catalogItem = this;
+
```
Comment (nf-s) [inline lib/ModelMixins/GeojsonMixin.ts]: This is overwriting the `feature` import from `"@turf/helpers"`

Can this be done generically at the bottom of `forceLoadMapItems`? - for example

```ts
        this._dataSource?.entities.values.map(
          entity => ((entity as any)._catalogItem = this)
        );
```

File: lib/ModelMixins/GeojsonMixin.ts [hunk 10]
```diff
@@ -997,6 +1019,9 @@ function GeoJsonMixin<T extends Constructor<Model<GeoJsonTraits>>>(Base: T) {
 
         const properties = entity.properties;
 
+        // _catalogItem property is needed for some feature picking functions (eg FeatureInfoMixin)
+        (entity as any)._catalogItem = this;
+
```
Comment (nf-s) [inline lib/ModelMixins/GeojsonMixin.ts]: See above - regarding doing the generically in `forceLoadMapItems`

File: lib/Models/Catalog/CatalogItems/GeoRssCatalogItem.ts [hunk 11]
```diff
@@ -178,7 +127,7 @@ class GeoRssStratum extends LoadableStratum(GeoRssCatalogItemTraits) {
 
 StratumOrder.addLoadStratum(GeoRssStratum.stratumName);
 
-export default class GeoRssCatalogItem extends MappableMixin(
+export default class GeoRssCatalogItem extends GeoJsonMixin(
   UrlMixin(CatalogMemberMixin(CreateModel(GeoRssCatalogItemTraits)))
 ) {
```
Comment (nf-s) [inline lib/Models/Catalog/CatalogItems/GeoRssCatalogItem.ts]: This should have  `implements HasLocalData`

**Review by nf-s — CHANGES_REQUESTED** [review]
Thanks for the PR @zoran995 !
Sorry it took so long to review

File: lib/Map/LeafletPatched.js (hunk 5 above)
Comment (zoran995) [reply lib/Map/LeafletPatched.js]: Initially, when I implemented this we were on 1.7 of Leaflet, in 1.8 fakeStop is removed and apparently not needed anymore as it is removed in the original 1.8 implementation
https://github.com/Leaflet/Leaflet/pull/7439/files#diff-f4127fe2c74bbf5c584e01246849fac1954ead8521449585b61d268bfb7b059eL358

File: lib/ModelMixins/FeatureInfoMixin.ts (hunk 6 above)
Comment (zoran995) [reply lib/ModelMixins/FeatureInfoMixin.ts]: thanks

**Comment by zoran995** [comment]
This is a new init file that can be used for testing (geojson, georss and czml)
https://gist.githubusercontent.com/zoran995/8200b51b38ff6c76a954b52b223e4319/raw/aacaec1bf485a281c42aa5aac8a2e372de4a8df3/feature-info-url-template-test-data-2.json

**Review by nf-s — APPROVED** [review]
Works well!

Thanks a lot @zoran995

**Pull request merged**